test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/TransactionTests.cs

```diff
@@ -3,35 +3,279 @@
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenAValidBundleWithForbiddenUser_WhenSubmittingATransaction_ThenOperationOutcomeWithForbiddenStatusIsReturned()
+        {
+            FhirClient tempClient = Client.CreateClientForUser(TestUsers.ReadOnlyUser, TestApplications.NativeClient);
+            var requestBundle = Samples.GetJsonSample("Bundle-TransactionWithValidBundleEntry");
+
+            var fhirException = await Assert.ThrowsAsync<FhirException>(async () => await tempClient.PostBundleAsync(requestBundle.ToPoco<Bundle>()));
+            Assert.Equal(HttpStatusCode.Forbidden, fhirException.StatusCode);
+
+            string[] expectedDiagnostics = { "Transaction failed on 'POST' for the requested url '/Patient'.", "Authorization failed." };
+            IssueType[] expectedCodeType = { OperationOutcome.IssueType.Processing, OperationOutcome.IssueType.Forbidden };
+            ValidateOperationOutcome(expectedDiagnostics, expectedCodeType, fhirException.OperationOutcome);
+        }
+
+        [Fact]
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenABundleWithInvalidConditionalReferenceInResourceBody_WhenSubmittingATransaction_ThenProperOperationOutComeIsReturned()
+        {
+            var observation = new Observation
+            {
+                Subject = new ResourceReference
+                {
+                    Reference = "Patient?identifier=http:/example.org/fhir/ids|234235",
+                },
+            };
+
+            var bundle = new Bundle
+            {
+                Type = BundleType.Transaction,
+                Entry = new List<EntryComponent>
+                {
+                    new EntryComponent
+                    {
+                        Resource = observation,
+                        Request = new RequestComponent
+                        {
+                            Method = HTTPVerb.POST,
+                            Url = "Observation",
+                        },
+                    },
+                },
+            };
+
+            var fhirException = await Assert.ThrowsAsync<FhirException>(async () => await Client.PostBundleAsync(bundle));
+
+            Assert.Equal(HttpStatusCode.BadRequest, fhirException.StatusCode);
+
+            string[] expectedDiagnostics = { "Given conditional reference 'Patient?identifier=http:/example.org/fhir/ids|234235' does not resolve to a resource." };
+            IssueType[] expectedCodeType = { IssueType.Invalid };
+            ValidateOperationOutcome(expectedDiagnostics, expectedCodeType, fhirException.OperationOutcome);
+        }
+
+        private static void ValidateOperationOutcome(string[] expectedDiagnostics, IssueType[] expectedCodeType, OperationOutcome operationOutcome)
+        {
+            Assert.NotNull(operationOutcome.Id);
+            Assert.NotEmpty(operationOutcome.Issue);
+
+            for (int iter = 0; iter < operationOutcome.Issue.Count; iter++)
```

Comment:
Should we assert that `expectedDiagnostics`, `expectedCodeType`, and `operationOutcome.Issue` all have the same length? We're sort of doing it here implicitly, but it might be something we want to be explicit about.